File: solutions/0032_longest_valid_parentheses.py
# Author: Xiangyue Cai
# Date: 2019-11-26

"""
Example 1:

Input: "(()"
Output: 2
Explanation: The longest valid parentheses substring is "()"
Example 2:

Input: ")()())"
Output: 4
Explanation: The longest valid parentheses substring is "()()"
"""

# runtime: 40ms beats 96.22%
# memory usage: 13.4mb beats 61.11%
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        res = 0
        last = -1
        stack = []
        if s is None or len(s) < 2:
            return res
        
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
            else:
                if stack == []:
                    last = i
                else:
                    stack.pop()
                    if stack == []:
                        res = max(res, i - last)
                    else:
                        res = max(res, i - stack[-1])
        return res

File: solutions/0032_longest_valid_parentheses.cc
/**
 * Author: Xiangyue Cai
 * Date: 2019-11-27
 */

// Example 1:

// Input: "(()"
// Output: 2
// Explanation: The longest valid parentheses substring is "()"
// Example 2:

// Input: ")()())"
// Output: 4
// Explanation: The longest valid parentheses substring is "()()"

// runtime 4ms
class Solution {
public:
    int longestValidParentheses(string s) {
        stack<int> st;
        st.push(-1);
        int max_len = 0;
        for (int i = 0 ; i < s.length() ; i++) {
            if (s[i] == '(') {
                st.push(i);
            } else if (s[i] == ')') {
                st.pop();
                if (st.empty())
                    st.push(i);
                max_len = max(max_len, i - st.top());
            }
        }
        return max_len;
    }
};